compiler/src/main/java/com/squareup/anvil/compiler/ModuleMerger.kt

```diff
@@ -92,141 +78,127 @@ internal class ModuleMerger(
         // modules contain the same bindings and are contributed to the same scope. To avoid this
         // issue we filter all generated Anvil modules except for the one that was generated for
         // this specific class.
-        val fqName = it.fqNameSafe
-        !fqName.isAnvilModule() || fqName == anvilModuleName
+        !it.fqName.isAnvilModule() || it.fqName == anvilModuleName
       }
-      .mapNotNull {
-        val contributesAnnotation =
-          it.annotationOrNull(contributesToFqName, scope = scopeFqName)
-            ?: return@mapNotNull null
-        it to contributesAnnotation
+      .mapNotNull { contributedClass ->
+        val contributesAnnotation = contributedClass.findAnnotations(
+          annotationName = contributesToFqName,
+          scopeName = mergeScope.fqName
+        ).firstOrNull() ?: return@mapNotNull null
+
+        contributesAnnotation
       }
-      .filter { (classDescriptor, _) ->
-        val moduleAnnotation = classDescriptor.annotationOrNull(daggerModuleFqName)
-        val mergeModulesAnnotation = classDescriptor.annotationOrNull(mergeModulesFqName)
-        if (!DescriptorUtils.isInterface(classDescriptor) &&
+      .filter { contributesAnnotation ->
+        val contributedClass = contributesAnnotation.declaringClass()
+        val moduleAnnotation = contributedClass.findAnnotations(daggerModuleFqName).firstOrNull()
+        val mergeModulesAnnotation =
+          contributedClass.findAnnotations(mergeModulesFqName).firstOrNull()
+
+        if (!contributedClass.isInterface() &&
           moduleAnnotation == null &&
           mergeModulesAnnotation == null
         ) {
-          throw AnvilCompilationException(
-            classDescriptor,
-            "${classDescriptor.fqNameSafe} is annotated with " +
+          throw AnvilCompilationExceptionClassReference(
+            contributedClass,
+            "${contributedClass.fqName} is annotated with " +
               "@${ContributesTo::class.simpleName}, but this class is neither an interface " +
               "nor a Dagger module. Did you forget to add @${Module::class.simpleName}?"
           )
         }
 
-        moduleAnnotation != null || mergeModulesAnnotation != null
-      }
-      .onEach { (classDescriptor, _) ->
-        if (classDescriptor.effectiveVisibility() !is Public) {
-          throw AnvilCompilationException(
-            classDescriptor,
-            "${classDescriptor.fqNameSafe} is contributed to the Dagger graph, but the " +
-              "module is not public. Only public modules are supported."
-          )
+        contributedClass.checkClassIsPublic {
+          "${contributedClass.fqName} is contributed to the Dagger graph, but the " +
+            "module is not public. Only public modules are supported."
         }
+
+        moduleAnnotation != null || mergeModulesAnnotation != null
       }
+      .cast<Sequence<AnnotationReference.Descriptor>>()
```

Comment:
This cast call is weird. You can avoid this by changing your extension functions to 
```kotlin
public fun <T : AnnotationReference> List<T>.findAnnotations(
  annotationName: FqName,
  scopeName: FqName? = null
): List<T> {
```
(I'd move it into `AnnotationReference` in this case)

Reply:
I tested it locally and it works. But there's another issue you need to fix for `ClassReference.Descriptor`:
```
    override val annotations: List<AnnotationReference.Descriptor> by lazy(NONE) {
      clazz.annotations.map { it.toAnnotationReference(this, module) }
    }
```
The current implementation returns just `AnnotationReference`.